### PR TITLE
RFC regarding the calculation of Base Versions

### LINKED ISSUE #2394: [Bug] GitVersionTask: VersionSource is set to wrong tag
**Describe the bug**
Using the default configuration of GitVersionTask v5.3.7 the VersionSource ends up being not the latest tagged commit on master but rather the one before that.

Image of the test repository branch history: (blue: develop, red: master, green: release branches)
image

Output of `gitversion /diag`:

````
INFO [08/26/20 16:21:25:50] Dumping commit graph: 
INFO [08/26/20 16:21:25:62] *   4c6828d 2 hours ago  (HEAD -> develop, origin/develop)
|\  
| *   27f83ef 2 hours ago  (tag: 0.4.0, origin/master, master)
| |\  
| | * 925521e 2 hours ago 
| |/  
|/|   
* | ac89157 2 hours ago 
* | b6e93d6 2 hours ago 
|\| 
| *   b6df2d3 2 hours ago  (tag: 0.3.0)
| |\  
| | * 27277e7 3 hours ago 
| |/  
|/|   
* | ae5d80a 3 hours ago 
* | 73ffb7d 3 hours ago 
|\| 
| *   a3fe19c 3 hours ago  (tag: 0.2.0)
| |\  
| | * 9a40911 3 hours ago 
| |/  
|/|   
* | c074d61 4 hours ago 
* | 00cdcbe 5 hours ago 
|/  
* 18dbaa2 5 hours ago  (tag: 0.1.0)

INFO [08/26/20 16:21:25:62] Working directory: D:\Workspace\GitVersionTest
INFO [08/26/20 16:21:25:67] Project root is: D:\Workspace\GitVersionTest\
INFO [08/26/20 16:21:25:67] DotGit directory is: D:\Workspace\GitVersionTest\.git
INFO [08/26/20 16:21:25:76] Using latest commit on specified branch
INFO [08/26/20 16:21:25:79] Running against branch: develop (4c6828daf637c00e93f7e1d7b47e029a0d867bd2)
INFO [08/26/20 16:21:25:79] Begin: Calculating base versions
  INFO [08/26/20 16:21:25:87] Found commit [4c6828daf637c00e93f7e1d7b47e029a0d867bd2] matching merge message format: Default
  INFO [08/26/20 16:21:25:87] Found commit [4c6828daf637c00e93f7e1d7b47e029a0d867bd2] matching merge message format: Default
  INFO [08/26/20 16:21:25:87] Found commit [4c6828daf637c00e93f7e1d7b47e029a0d867bd2] matching merge message format: Default
  INFO [08/26/20 16:21:25:87] Merge message 'Merge branch 'release/0.4.0'': 0.4.0 with commit count source 27f83ef810557a4f92a310c5ecd3d37e946a9c50
  INFO [08/26/20 16:21:25:88] Merge message 'Merge branch 'release/0.3.0'': 0.3.0 with commit count source b6df2d3925f42079b32ba28321368f4905159926
  INFO [08/26/20 16:21:25:88] Merge message 'Merge branch 'release/0.2.0'': 0.2.0 with commit count source a3fe19c9fcbc8e3b32238c31e321e8b2bc800b57
  INFO [08/26/20 16:21:25:89] Git tag '0.4.0': 0.4.0 with commit count source 27f83ef810557a4f92a310c5ecd3d37e946a9c50
  INFO [08/26/20 16:21:25:89] Git tag '0.3.0': 0.3.0 with commit count source b6df2d3925f42079b32ba28321368f4905159926
  INFO [08/26/20 16:21:25:89] Git tag '0.2.0': 0.2.0 with commit count source a3fe19c9fcbc8e3b32238c31e321e8b2bc800b57
  INFO [08/26/20 16:21:25:89] Git tag '0.1.0': 0.1.0 with commit count source 18dbaa2bda565ca554463be8f7797343501cb0d3
  INFO [08/26/20 16:21:25:90] Git tag '0.4.0': 0.4.0 with commit count source 27f83ef810557a4f92a310c5ecd3d37e946a9c50
  INFO [08/26/20 16:21:25:91] Git tag '0.3.0': 0.3.0 with commit count source b6df2d3925f42079b32ba28321368f4905159926
  INFO [08/26/20 16:21:25:91] Git tag '0.2.0': 0.2.0 with commit count source a3fe19c9fcbc8e3b32238c31e321e8b2bc800b57
  INFO [08/26/20 16:21:25:91] Git tag '0.1.0': 0.1.0 with commit count source 18dbaa2bda565ca554463be8f7797343501cb0d3
  INFO [08/26/20 16:21:25:92] Found multiple base versions which will produce the same SemVer (0.5.0), taking oldest source for commit counting (Merge message 'Merge branch 'release/0.4.0'')
  INFO [08/26/20 16:21:25:92] Base version used: Merge message 'Merge branch 'release/0.4.0'': 0.4.0 with commit count source b6df2d3925f42079b32ba28321368f4905159926
  INFO [08/26/20 16:21:25:92] End: Calculating base versions (Took: 126.65ms)
  INFO [08/26/20 16:21:25:92] 5 commits found between b6df2d3925f42079b32ba28321368f4905159926 and 4c6828daf637c00e93f7e1d7b47e029a0d867bd2
  INFO [08/26/20 16:21:25:93] Begin: Getting version tags from branch 'refs/heads/develop'.
  INFO [08/26/20 16:21:25:94] End: Getting version tags from branch 'refs/heads/develop'. (Took: 9.33ms)
{
  "Major":0,
  "Minor":5,
  "Patch":0,
  "PreReleaseTag":"alpha.5",
  "PreReleaseTagWithDash":"-alpha.5",
  "PreReleaseLabel":"alpha",
  "PreReleaseNumber":5,
  "WeightedPreReleaseNumber":5,
  "BuildMetaData":"",
  "BuildMetaDataPadded":"",
  "FullBuildMetaData":"Branch.develop.Sha.4c6828daf637c00e93f7e1d7b47e029a0d867bd2",
  "MajorMinorPatch":"0.5.0",
  "SemVer":"0.5.0-alpha.5",
  "LegacySemVer":"0.5.0-alpha5",
  "LegacySemVerPadded":"0.5.0-alpha0005",
  "AssemblySemVer":"0.5.0.0",
  "AssemblySemFileVer":"0.5.0.0",
  "FullSemVer":"0.5.0-alpha.5",
  "InformationalVersion":"0.5.0-alpha.5+Branch.develop.Sha.4c6828daf637c00e93f7e1d7b47e029a0d867bd2",
  "BranchName":"develop",
  "EscapedBranchName":"develop",
  "Sha":"4c6828daf637c00e93f7e1d7b47e029a0d867bd2",
  "ShortSha":"4c6828d",
  "NuGetVersionV2":"0.5.0-alpha0005",
  "NuGetVersion":"0.5.0-alpha0005",
  "NuGetPreReleaseTagV2":"alpha0005",
  "NuGetPreReleaseTag":"alpha0005",
  "VersionSourceSha":"b6df2d3925f42079b32ba28321368f4905159926",
  "CommitsSinceVersionSource":5,
  "CommitsSinceVersionSourcePadded":"0005",
  "CommitDate":"2020-08-26"
}
  INFO [08/26/20 16:21:25:97] Done writing 

````

## Expected Behavior
I would expect the VersionSource to be the 0.4.0 tag

## Actual Behavior
VersionSource is determined to be the 0.3.0 tag

I tried my hand on this fixture thing but i have no idea if i did it right:

```
[Test]
    public void VersionSource()
    {
        // Arrange
        var config = new ConfigurationBuilder().Build();

        // Act
        using var fixture = new BaseGitFlowRepositoryFixture("0.1.0");
        fixture.Checkout("master");
        var initialCommit = fixture.Repository.Head.Tip;
        fixture.MergeNoFF("develop");
        fixture.Checkout("develop");
        fixture.MakeACommit("Feature commit 1");
        fixture.BranchTo("release/0.2.0");
        fixture.MakeACommit("Release commit 1");
        fixture.Checkout("master");
        fixture.MergeNoFF("release/0.2.0");
        fixture.ApplyTag("0.2.0");
        var tag = fixture.Repository.Head.Tip;
        fixture.Checkout("develop");
        fixture.MergeNoFF("master");
        var version = fixture.GetVersion(config);
        version.VersionSourceSha.ShouldBe(tag.Sha);
    }
```

- Version Used:
- Operating System and version (Windows 10, Ubuntu 18.04): Windows 10 Version 10.0.17763 Build 17763

### LINKED ISSUE #2798: [Improvement] RFC regarding the calculation of Base Versions
## Description

I was facing a problem with GitVersion today (where the presence/absence of a release made the version of another branch jump around like crazy) and after some investigation, I understand that the resolution of  a `BaseVersionSource` of a `MergeMessage` depends on the presence/absence of a release in the repository. (`FixTheBaseVersionSourceOfMergeMessageStrategyIfReleaseBranchWasMergedAndDeleted`)

That code was added to stop versions from jumping around when a release branch is deleted. (https://github.com/GitTools/GitVersion/issues/1526)

But when I investigated further, I started wondering why we were not using the `BaseVersionSource` from `Tags` first, before relying on the `BaseVersionSource` if `MergeMessage`. (this is actually how I understood this doc at first : https://gitversion.net/docs/reference/version-sources)

## Possible Implementation

This PR does just that and I would greatly appreciate your opinion about it, as there could be unforeseen drawbacks. : https://github.com/GitTools/GitVersion/pull/2799

## Drawbacks

As you'll be able to see in the unit tests, the version is calculated differently when merge commits are involved. e.g. if `develop` is at `1.1.2-alpha.2` and we merge in a `feature/*` branch with 10 commits, the new resulting version would be `1.1.2-alpha.3` instead of `1.1.2-alpha.12` (or 13), as the MergeCommit is counted as 1 commit only.

This the version always keeps increasing, so all is good on that side, but I fear that the projects that are already at higher versions would be greatly impacted, when their version would be sent backwards on a GitVersion update.

Thanks a lot for taking the time to read this!

### PR DESCRIPTION
**Pull request by franknarf8**
Dear Reviewers,

Please take a look at the discussion here : https://github.com/GitTools/GitVersion/issues/2798

TL;DR : I'm proposing we use the `BaseVersionSource` of `Tags` before `MergeMessage`, if there are any.

I think there might be implications I am not foreseeing and would like your input on this proposition.

## Related Issue

Resolves #2798.
Related to #1526.
Fixes #2394.

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs
Comment (franknarf8) [inline src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs:63]: I've also improved the logging here that I thought was misleading. It was initially saying that there were many base versions, when there were only 1.

So, if we don't get the logic change, I'll refactor this PR just to get the better logging in.

**Comment by asbjornu** [comment]
I think this is an improvement, but as it is a breaking change, I think we need to postpone this to GitVersion 6.

**Comment by ni-fgenois** [comment]
That's a fair point; so when is v6 planned for? And what are the next steps to get this in?

Also, I'm having some trouble fixing the last 2 failing tests, could someone have a look?

**Comment by asbjornu** [comment]
I think that if you rebase on `main`, a new run of the tests should turn them green.

**Comment by ni-fgenois** [comment]
Still seems broken for some other reason :S

**Comment by asbjornu** [comment]
Just a note for your next PR: Create a disposable branch from `main` and do your commits in it and submit your pull request based on that branch instead of using `main`, since `main` should preferably be identical across forks to make submitting new pull requests easier.

**Comment by ni-fgenois** [comment]
I don't mind fixing this merge conflict again, but will it be merged afterwards? Does it need more approvals?

**Comment by asbjornu** [comment]
@ni-fgenois, sorry about not merging this, but as mentioned in https://github.com/GitTools/GitVersion/pull/2799#issuecomment-893050025 we need to postpone this to version 6 of GitVersion as it is a breaking change. We don't have a v6 branch yet, but perhaps we should create one so we can merge v6 features like this. Thoughts @arturcic?

**Comment by arturcic** [comment]
@asbjornu I would go the other way, I'd create a `v5.x` branch and `main` would be for the `v6`. But first I want to finish with some warnings fixing. Version 5.x will be only for bug fixing and no new features will be added.

**Comment by ni-fgenois** [comment]
Ok, please tell me when the `5.x` release will be kicked, and I'll gladly fix the merge conflict :)

Thank you two for the quick response :)

**Comment by asbjornu** [comment]
Sounds like a plan, @arturcic! 👍🏼

**Comment by ni-fgenois** [comment]
The 5.x release seem to have been kicked, is it time to merge this in? (If so, I'll rebase the code and fix the tests)

**Comment by arturcic** [comment]
@ni-fgenois not yet, I'm working on making the branch `release\5.x` release-able from the build infrastructure, and want to enable the nullable for the entire solution. After that we'll focus on `main` for v6. For now `main` and `release\5.x` will be the same

**Comment by asbjornu** [comment]
`main` is now primed for v6. Can you please rebase and solve conflicts so we can merge this, @franknarf8?

**Comment by ni-fgenois** [comment]
Sorry for the delay, the merge conflict was a bit more challenging this time around.

You might want to also take a look to this new test case I have added here : https://github.com/GitTools/GitVersion/pull/2799/files#diff-d43fa8360d2034a4a95db32f5376f489764f0cb4b4ceec376940a7797a699528R205

Thanks for the notification and all the support :)

**Comment by asbjornu** [comment]
Awesome, @ni-fgenois. Thanks! Just one thing I thought when reviewing this is that there has been some discussion leading to #3041 lately. I think this PR affects that and want to have your thoughts on the matter.

**Comment by ni-fgenois** [comment]
This PR would indeed partially impact that other topic; as we would now be taking versioned tags over any other VersionSources (merge commits, etc). But if there is a case where there are no tags present in the VersionSources, the old behavior will still apply.

This relates to another interrogation I had on this topic the other day about the behavior of GitVersion when encoutering tags. I thought that, running GitVersion on a commit with a tag that is a valid SemVer should simply return that version (even when there is some Meta information in the tag). What do you think? I thought this would be a convenient way to lock/override a version (helps with replayability of GitVersion).

This came to me when dealing with tags on master containing Release Candidate meta information. I wanted to setup automatic tagging on `master` while using Continuous Delivery, but GitVersion infers a different version before and after tagging a version with RC metadata. For example, when running GitVersion a first time results in `2.0.0+0`, but adding the `2.0.0+0` tag on that same commit causes the next version to be `2.0.0` instead of staying at `2.0.0+0`. I guess we could also reload the `CommitCountSource` from those tags (if present) and count from there, what do you think?

Here's a test that illustrates this scenario : 

```
[Test]
public void TestTagsWithMetadataInfo()
{
    using var fixture = new EmptyRepositoryFixture();
    fixture.Repository.MakeATaggedCommit("1.0.0");
    fixture.Repository.MakeCommits(1);
    fixture.Repository.CreateBranch("release/2.0.0");
    fixture.Checkout("release/2.0.0");
    fixture.Repository.MakeCommits(4);
    fixture.Checkout(MainBranch);
    fixture.Repository.MergeNoFF("release/2.0.0", Generate.SignatureNow());

    fixture.AssertFullSemver("2.0.0+0");
    fixture.Repository.MakeCommits(1);
    fixture.AssertFullSemver("2.0.0+1");
    fixture.Repository.ApplyTag("2.0.0+1");
    fixture.AssertFullSemver("2.0.0"); // instead of 2.0.0+1
    fixture.Repository.MakeCommits(1);
    fixture.AssertFullSemver("2.0.1+1"); // instead of 2.0.0+2
}
```

**Comment by ni-fgenois** [comment]
So, I've updated the failing test specified in https://github.com/GitTools/GitVersion/issues/2394 to use the new fixtures (see below), and it seems my PR fixes this behavior (the test below is green on my branch; haven't pushed it though).

I haven't found a test case to repro the issue mentioned in : https://github.com/GitTools/GitVersion/issues/2394

```c#
[Test]
    public void VersionSource()
    {
        using var fixture = new EmptyRepositoryFixture();
        fixture.Repository.MakeATaggedCommit("0.1.0");
        fixture.Checkout(MainBranch);
        fixture.Repository.CreateBranch("develop");
        fixture.Checkout("develop");
        fixture.MakeACommit("Feature commit 1");
        fixture.BranchTo("release/0.2.0");
        fixture.MakeACommit("Release commit 1");
        fixture.Checkout(MainBranch);
        fixture.MergeNoFF("release/0.2.0");
        fixture.ApplyTag("0.2.0");
        var tag = fixture.Repository.Head.Tip;
        fixture.Checkout("develop");
        fixture.MergeNoFF(MainBranch);
        fixture.AssertFullSemver("0.3.0-alpha.1");
        var version = fixture.GetVersion();
        version.VersionSourceSha.ShouldBeEquivalentTo(tag.Sha);
    }
```

**Comment by asbjornu** [comment]
That sounds great, @ni-fgenois! Please include the test from #2394 in this PR and write in the description `Fixes #2394` so it'll be closed when this is merged.

**Comment by ni-fgenois** [comment]
Done! :)

On another subject, while I've got you here, what do you think of this behavior illustrated above?

```c#
fixture.AssertFullSemver("2.0.0+1");
fixture.Repository.ApplyTag("2.0.0+1");
fixture.AssertFullSemver("2.0.0"); // shouldn't it be 2.0.0+1?
```

Is there already an explaination for that, or should I open another ticket?

**Comment by asbjornu** [comment]
> Done! :)

Great!

> On another subject, while I've got you here, what do you think of this behavior illustrated above?
> 
> ```cs
> fixture.AssertFullSemver("2.0.0+1");
> fixture.Repository.ApplyTag("2.0.0+1");
> fixture.AssertFullSemver("2.0.0"); // shouldn't it be 2.0.0+1?
> ```
> 
> Is there already an explaination for that, or should I open another ticket?

I'm not sure there is a good explanation other than that everything after `+` is just metadata and something that is provided by GitVersion, but not something that GitVersion sources from somewhere else. So when GitVersion reads the tag `2.0.0+1` it just ignores `+1`. Then, when the `FullSemVer` is generated, the result is `2.0.0` since this is a stable version provided by a tag. 

Although I don't think careful consideration has been applied to this behavior, I would dare to say that this is by design and not a bug. Whether the design is correct or not, can of course be discussed.

File: src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs
Comment (asbjornu) [inline src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs:124]:
```suggestion
            this.log.Info($"Base version used: {calculatedBase}");
```

File: src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs
Comment (asbjornu) [inline src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs:98]:
```suggestion
                    this.log.Info(
```

File: src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs
Comment (asbjornu) [inline src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs:92]:
```suggestion
                    this.log.Info($"Taking oldest source for commit counting : {BaseVersionToString(baseVersionWithOldestSource)}");
```

File: src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs
Comment (asbjornu) [inline src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs:86]:
```suggestion
                        this.log.Info("As there are Git tags, the other sources will be discarded");
```

File: src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs
Comment (asbjornu) [inline src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs:78]:
```suggestion
                            this.log.Info($" - {BaseVersionToString(baseVersion)}");
```

File: src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs
Comment (asbjornu) [inline src/GitVersion.Core/VersionCalculation/BaseVersionCalculator.cs:73]:
```suggestion
                    this.log.Info($"Found multiple base versions which will produce the same SemVer ({maxVersion.IncrementedVersion})");
                    this.log.Info($"Here are the different source candidate for commit counting : ");
```

**Review by asbjornu — CHANGES_REQUESTED** [review]
The only thing to point out are some minor code style changes. Also, I think it would be great to document how this works once and for all. Could you please attempt to write something up? I'll help with the formulation, but I would appreciate if you could get the ball rolling. 🙏🏼

**Comment by asbjornu** [comment]
I believe #3190 touched upon many of the same files as this PR. Can you please rebase and assess whether this PR is relevant and needed anymore?

**Comment by ni-fgenois** [comment]
To be honest, I kinda lost track of this change over the last couple of months. I'm not even sure if this PR is relevant anymore; I would suggest closing it for now. I can always re-open it if needed.

**Comment by HHobeck** [comment]
Hmm I see... With the refactoring of https://github.com/GitTools/GitVersion/pull/3190 the BaseVersionCalculator class has not been survived. That means your refactoring is obsolete IMO. Sorry for that! But if I see your unit tests then this is almost fixed in 6.x and the behavior is like you would expecting it.

**Comment by HHobeck** [comment]
> On another subject, while I've got you here, what do you think of this behavior illustrated above?
> 
> ```cs
> fixture.AssertFullSemver("2.0.0+1");
> fixture.Repository.ApplyTag("2.0.0+1");
> fixture.AssertFullSemver("2.0.0"); // shouldn't it be 2.0.0+1?
> ```
> 
> Is there already an explaination for that, or should I open another ticket?

I totally agree with you that the output of GitVersion should yield the same result before and after the tagging has been done. I would go one step further and say: Even if you tag the commit with 2.0.0.

**Comment by HHobeck** [comment]
Please help us to realize https://github.com/GitTools/GitVersion/issues/3041 and contribute for version 6.x. I'm going to close this PR because it makes no sense anymore. Thank you very very very much for your input and your work.

**Pull request closed without merging**
